### PR TITLE
CDF-22143: add workflow trigger api support

### PR DESCRIPTION
**Pull request by vincent-cognite**
## Description
Add suuport for Workflow Triggers api
Tested with the current deployment in greenfield. Some minor renamings are expected.

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/_api/workflows.py [hunk 1]
```diff
@@ -47,6 +53,130 @@ def wrap_workflow_ids(
...
+                ...         external_id="my_trigger",
+                ...         trigger_rule=WorkflowScheduledTriggerRule(cron_spec="0 0 * * *"),
+                ...         workflow_external_id="my_workflow",
+                ...         workflow_version="1",
+                ...         input_data={"a": 1, "b": 2},
+                ...     )
+                ... )
+        """
+        if workflow_trigger.authentication is None:
+            nonce = create_session_and_return_nonce(
```
Comment (VerstraeteBert) [inline cognite/client/_api/workflows.py:105]:
```suggestion
        if workflow_trigger.authentication is None:
            nonce = create_session_and_return_nonce(
```
Let's forego allowing setting a nonce directly similar to the .trigger() endpoint

Will need a WorkflowTriggerRead and Write model. I think that's good practice anyways, less confusion with the fields that are set by our API (lastUpdated, ...)

File: cognite/client/data_classes/workflows.py [hunk 2]
```diff
@@ -1175,3 +1175,146 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+    """
+    This class represents a scheduled trigger rule.
+
+    Args:
+        cron_spec (str | None): The cron specification for the scheduled trigger.
+    """
+
+    def __init__(self, cron_spec: str | None = None) -> None:
+        self.type: WorkflowTriggerType = "scheduledTrigger"
+        # self.trigger_type: WorkflowTriggerType = "scheduledTrigger"
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1193]:
```suggestion
```

File: cognite/client/data_classes/workflows.py [hunk 3]
```diff
@@ -1175,3 +1175,146 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+
+WorkflowTriggerType: TypeAlias = Literal["scheduledTrigger"]
+
+
+class WorkflowScheduledTriggerRule(CogniteObject):
+    """
+    This class represents a scheduled trigger rule.
+
+    Args:
+        cron_spec (str | None): The cron specification for the scheduled trigger.
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1188]: It's cron_expression now

File: cognite/client/data_classes/workflows.py [hunk 4]
```diff
@@ -1175,3 +1175,146 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+class WorkflowTrigger(CogniteResource):
+    """
+    This class represents a workflow trigger.
+
+    Args:
+        external_id (str): The external ID provided by the client. Must be unique for the resource type.
+        trigger_rule (WorkflowScheduledTriggerRule): The trigger rule of the workflow version trigger.
+        workflow_external_id (str): The external ID of the workflow.
+        workflow_version (str): The version of the workflow.
+        input_data (dict | None): The input data of the workflow version trigger. Defaults to None.
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1206]: It's input now

File: cognite/client/data_classes/workflows.py [hunk 5]
```diff
@@ -1175,3 +1175,146 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+        cron_spec (str | None): The cron specification for the scheduled trigger.
+    """
+
+    def __init__(self, cron_spec: str | None = None) -> None:
+        self.type: WorkflowTriggerType = "scheduledTrigger"
+        # self.trigger_type: WorkflowTriggerType = "scheduledTrigger"
+        self.cron_spec = cron_spec
+
+
+class WorkflowTrigger(CogniteResource):
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1197]: I'm not a big fan of the shared Read and Write class, may be confusing for users of the create endpoint.

File: cognite/client/data_classes/workflows.py [hunk 6]
```diff
@@ -1175,3 +1175,146 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
 
     def dump(self, camel_case: bool = True, as_external_id: bool = False) -> list[dict[str, Any]]:
         return [workflow_id.dump(camel_case, as_external_id_key=as_external_id) for workflow_id in self.data]
+
+
+WorkflowTriggerType: TypeAlias = Literal["scheduledTrigger"]
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1180]: it's 'schedule' now

File: cognite/client/data_classes/workflows.py [hunk 7]
```diff
@@ -1175,3 +1175,146 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+
+class WorkflowTriggerRun(CogniteResource):
+    """
+    This class represents a workflow trigger run.
+    """
+
+    def __init__(
+        self,
+        trigger_external_id: str,
+        trigger_type: WorkflowTriggerType,
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1281]: triggerRun does not contain a trigger type anymore

File: cognite/client/_api/workflows.py [hunk 8]
```diff
@@ -72,32 +73,32 @@ class WorkflowTriggerAPI(BetaWorkflowAPIClient):
...
         self,
-        workflow_trigger: WorkflowTrigger,
+        workflow_trigger: WorkflowTriggerCreate,
         client_credentials: ClientCredentials | None = None,
     ) -> WorkflowTrigger:
         """`Create a new trigger for a workflow. <https://api-docs.cognite.com/20230101/tag/Workflow-triggers/operation/createTriggers>`_
 
         Args:
-            workflow_trigger (WorkflowTrigger): The workflow trigger specitification.
+            workflow_trigger (WorkflowTriggerWrite): The workflow trigger specitification.
```
Comment (VerstraeteBert) [inline cognite/client/_api/workflows.py]: WorkflowTriggerWrite and Create are both used here

File: cognite/client/data_classes/workflows.py [hunk 9]
```diff
@@ -1177,21 +1177,76 @@ def dump(self, camel_case: bool = True, as_external_id: bool = False) -> list[di
...
+    """
+    This class represents a workflow trigger for creation.
+
+    Args:
+        external_id (str): The external ID provided by the client. Must be unique for the resource type.
+        trigger_rule (WorkflowScheduledTriggerRule): The trigger rule of the workflow version trigger.
+        workflow_external_id (str): The external ID of the workflow.
+        workflow_version (str): The version of the workflow.
+        input (dict | None): The input data of the workflow version trigger. Defaults to None.
+        authentication (dict | None): The authentication of the workflow version trigger. Defaults to None.
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py]: The descriptions are not very useful. For example, for input, we could say "this object will be the workflow input to any execution started by the trigger".

**Review by VerstraeteBert — APPROVED** [review]

**Review by VerstraeteBert — APPROVED** [review]

File: tests/tests_unit/test_testing.py [hunk 10]
```diff
@@ -12,7 +13,7 @@
 def test_ensure_all_apis_are_available_on_cognite_mock():
     mocked_apis = all_mock_children(CogniteClientMock())
     available = {v.__class__ for v in mocked_apis.values()}
-    expected = set(all_subclasses(APIClient))
+    expected = set(all_subclasses(APIClient)) - {BetaWorkflowAPIClient}
```
Comment (doctrino) [inline tests/tests_unit/test_testing.py:16]: Why this exception? Why is it difficult to add to the mock client?

File: cognite/client/_api/workflows.py [hunk 11]
```diff
@@ -47,6 +54,130 @@ def wrap_workflow_ids(
...
+    ) -> None:
+        super().__init__(config, api_version, cognite_client)
+        self._api_subversion = "beta"
+        self._warning = FeaturePreviewWarning(
+            api_maturity="beta", sdk_maturity="beta", feature_name="Workflow Orchestration"
+        )
+
+
+class WorkflowTriggerAPI(BetaWorkflowAPIClient):
+    _RESOURCE_PATH = "/workflows/triggers"
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:72]: Why this parent class here? Why just add it to the `WorkflowTriggerAPI`? Then you do not have to make any exception in the mock test.

```suggestion
class WorkflowTriggerAPI(APIClient):
    _RESOURCE_PATH = "/workflows/triggers"

    def __init__(
        self,
        config: ClientConfig,
        api_version: str | None,
        cognite_client: CogniteClient,
    ) -> None:
        super().__init__(config, api_version, cognite_client)
        self._api_subversion = "beta"
        self._warning = FeaturePreviewWarning(
            api_maturity="beta", sdk_maturity="beta", feature_name="Workflow Orchestration"
        )
```

File: cognite/client/data_classes/workflows.py [hunk 12]
```diff
@@ -1175,3 +1175,185 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+        input (dict | None): The input data of the workflow version trigger. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        external_id: str,
+        trigger_rule: WorkflowScheduledTriggerRule,
+        workflow_external_id: str,
+        workflow_version: str,
+        input: dict | None = None,
```
Comment (doctrino) [inline cognite/client/data_classes/workflows.py:1214]: In the spec, `input` is listed as required. I assume that is a mistake and it should be optional?

File: cognite/client/data_classes/workflows.py [hunk 13]
```diff
@@ -1175,3 +1175,185 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+    Args:
+        cron_expression(str | None): The cron specification for the scheduled trigger.
+    """
+
+    def __init__(self, cron_expression: str | None = None) -> None:
+        self.trigger_type: WorkflowTriggerType = "schedule"
+        self.cron_expression = cron_expression
+
+
+class WorkflowTriggerCreate(CogniteResource):
```
Comment (doctrino) [inline cognite/client/data_classes/workflows.py:1196]: More a personal preference, but typically we have a `WorkflowTriggerCore`, which contains the shared parameter for the read and write format of the workflow trigger, that both `WorkflowTriggerCreate` and `WorkflowTrigger` subclasses.

File: cognite/client/data_classes/workflows.py [hunk 14]
```diff
@@ -1175,3 +1175,185 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+    """
+    This class represents a scheduled trigger rule.
+
+    Args:
+        cron_expression(str | None): The cron specification for the scheduled trigger.
+    """
+
+    def __init__(self, cron_expression: str | None = None) -> None:
+        self.trigger_type: WorkflowTriggerType = "schedule"
+        self.cron_expression = cron_expression
```
Comment (doctrino) [inline cognite/client/data_classes/workflows.py:1193]: I assume this is the first of many Trigger rules. Then the pattern we typically use is something like this 

(You can for example look at `PropertyType` in data modeling for a similar case`.

```suggestion
class WorkflowTriggerRule(CogniteObject, ABC):
    """This is the base class for all workflow trigger rules."""
    _trigger_type: ClassVar[str]

    @classmethod
    def _load(cls, resource: dict, cognite_client: CogniteClient | None = None) -> Self:
        trigger_type = resource["type"]
        if trigger_type in _TRIGGER_RULE_BY_TYPE:
            return cast(Self, _TRIGGER_RULE_BY_TYPE[trigger_type]._load_trigger(resource))
        # If more triggers are added in the future, this ensures that the SDK does not break.
        return cast(Self, UnknownCogniteObject(resource))

    @classmethod
    @abstractmethod
    def _load_trigger(cls, data: dict) -> Self:
        raise NotImplementedError()


class WorkflowScheduledTriggerRule(WorkflowTriggerRule):
    """
    This class represents a scheduled trigger rule.

    Args:
        cron_expression(str | None): The cron specification for the scheduled trigger.
    """
    _trigger_type = "schedule"

    def __init__(self, cron_expression: str | None = None) -> None:
        self.cron_expression = cron_expression

    @classmethod
    def _load_trigger(cls, data: dict) -> WorkflowScheduledTriggerRule:
        return cls(cron_expression=data.get("cronExpression"))


_TRIGGER_RULE_BY_TYPE: dict[str, type[WorkflowTriggerRule]] = {subclass._trigger_type: subclass for subclass in WorkflowTriggerRule.__subclasses__()}
```

File: cognite/client/data_classes/workflows.py [hunk 15]
```diff
@@ -1175,3 +1175,185 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+            return convert_all_keys_to_camel_case(item)
+        return item
+
+    @classmethod
+    def _load(cls, resource: dict, cognite_client: CogniteClient | None = None) -> WorkflowTriggerCreate:
+        return cls(
+            external_id=resource["externalId"],
+            workflow_external_id=resource["workflowExternalId"],
+            workflow_version=resource["workflowVersion"],
+            trigger_rule=WorkflowScheduledTriggerRule._load(resource["triggerRule"]),
```
Comment (doctrino) [inline cognite/client/data_classes/workflows.py:1241]:
```suggestion
            trigger_rule=WorkflowTriggerRule._load(resource["triggerRule"]),
```

Comment (doctrino) [reply cognite/client/data_classes/workflows.py:1241]: Important to not break the SDK when new triggers are added to the API

File: cognite/client/data_classes/workflows.py [hunk 16]
```diff
@@ -1175,3 +1175,185 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+            return convert_all_keys_to_camel_case(item)
+        return item
+
+    @classmethod
+    def _load(cls, resource: dict, cognite_client: CogniteClient | None = None) -> WorkflowTrigger:
+        return cls(
+            external_id=resource["externalId"],
+            workflow_external_id=resource["workflowExternalId"],
+            workflow_version=resource["workflowVersion"],
+            trigger_rule=WorkflowScheduledTriggerRule._load(resource["triggerRule"]),
```
Comment (doctrino) [inline cognite/client/data_classes/workflows.py:1301]:
```suggestion
            trigger_rule=WorkflowTriggerRule._load(resource["triggerRule"]),
```

File: cognite/client/_api/workflows.py [hunk 17]
```diff
@@ -47,6 +54,130 @@ def wrap_workflow_ids(
...
+        self,
+        limit: int = DEFAULT_LIMIT_READ,
+    ) -> WorkflowTriggerList:
+        """`Retrieve the trigger list. <https://api-docs.cognite.com/20230101/tag/Workflow-triggers/operation/getTriggers>`_
+
+        Args:
+            limit (int): Maximum number of results to return. Defaults to 25. Set to -1, float("inf") or None to return all items.
+
+        Returns:
+            WorkflowTriggerList: The trigger list.
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:148]: Example?

File: cognite/client/_api/workflows.py [hunk 18]
```diff
@@ -47,6 +54,130 @@ def wrap_workflow_ids(
...
+        limit: int = DEFAULT_LIMIT_READ,
+    ) -> WorkflowTriggerRunList:
+        """`List the history of runs for a trigger. <https://api-docs.cognite.com/20230101/tag/Workflow-triggers/operation/listTriggerRuns>`_
+
+        Args:
+            external_id (str): The external id of the trigger to list runs for.
+            limit (int): Maximum number of results to return. Defaults to 25. Set to -1, float("inf") or None to return all items.
+
+        Returns:
+            WorkflowTriggerRunList: The requested trigger runs.
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:170]: Example?

**Review by doctrino — CHANGES_REQUESTED** [review]
A few changes to the endpoints as well as some suggestion to follow the patterns

File: cognite/client/data_classes/workflows.py [hunk 19]
```diff
@@ -1175,3 +1175,185 @@ def load(cls, resource: Any, cognite_client: CogniteClient | None = None) -> Wor
...
+    """
+    This class represents a workflow trigger for creation.
+
+    Args:
+        external_id (str): The external ID provided by the client. Must be unique for the resource type.
+        trigger_rule (WorkflowScheduledTriggerRule): The trigger rule of the workflow version trigger.
+        workflow_external_id (str): The external ID of the workflow.
+        workflow_version (str): The version of the workflow.
+        input (dict | None): The input data of the workflow version trigger. Defaults to None.
+    """
```
[the remark this answers is not in the document]
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py:1206]: doc updated

File: tests/tests_unit/test_testing.py (hunk 10 above)
Comment (vincent-cognite) [reply tests/tests_unit/test_testing.py:16]: I followed what another PR had been doing. Thought it was the way to introduce beta methods.
I'll update.

File: cognite/client/data_classes/workflows.py (hunk 12 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py:1214]: It was a mistake indeed. We have updated the API spec and the server side too.

File: cognite/client/_api/workflows.py (hunk 17 above)
Comment (vincent-cognite) [reply cognite/client/_api/workflows.py:146]: example added

File: cognite/client/_api/workflows.py (hunk 18 above)
Comment (vincent-cognite) [reply cognite/client/_api/workflows.py:168]: example added

File: cognite/client/data_classes/workflows.py (hunk 13 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py]: I applied the same pattern too.

File: cognite/client/_api/workflows.py [hunk 20]
```diff
@@ -48,6 +54,144 @@ def wrap_workflow_ids(
...
+                ...     WorkflowTriggerCreate(
+                ...         external_id="my_trigger",
+                ...         trigger_rule=WorkflowScheduledTriggerRule(cron_expression="0 0 * * *"),
+                ...         workflow_external_id="my_workflow",
+                ...         workflow_version="1",
+                ...         input={"a": 1, "b": 2},
+                ...     )
+                ... )
+        """
+        nonce = create_session_and_return_nonce(
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:103]: Creating the warning is not enough :) 
```suggestion
        """
        self._warning.warn()
        nonce = create_session_and_return_nonce(
```

File: cognite/client/_api/workflows.py [hunk 21]
```diff
@@ -48,6 +54,144 @@ def wrap_workflow_ids(
...
+            external_id (str): The external id of the trigger to delete.
+
+        Examples:
+
+            Delete a trigger with external id 'my_trigger':
+
+                >>> from cognite.client import CogniteClient
+                >>> client = CogniteClient()
+                >>> client.workflows.triggers.delete("my_trigger")
+        """
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:130]:
```suggestion
        """
        self._warning.warn()
```

File: cognite/client/_api/workflows.py [hunk 22]
```diff
@@ -48,6 +54,144 @@ def wrap_workflow_ids(
...
+            WorkflowTriggerList: The trigger list.
+
+        Examples:
+
+            Get all triggers:
+
+                >>> from cognite.client import CogniteClient
+                >>> client = CogniteClient()
+                >>> res = client.workflows.triggers.get_triggers()
+        """
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:155]:
```suggestion
        """
        self._warning.warn()
```

File: cognite/client/_api/workflows.py [hunk 23]
```diff
@@ -48,6 +54,144 @@ def wrap_workflow_ids(
...
+            WorkflowTriggerRunList: The requested trigger runs.
+
+        Examples:
+
+            Get all runs for a trigger with external id 'my_trigger':
+
+                >>> from cognite.client import CogniteClient
+                >>> client = CogniteClient()
+                >>> res = client.workflows.triggers.get_trigger_run_history("my_trigger")
+        """
```
Comment (doctrino) [inline cognite/client/_api/workflows.py:185]:
```suggestion
        """
        self._warning.warn()
```

**Review by doctrino — APPROVED** [review]
Pre-approving, only need to trigger the feature preview warning when the user uses any of the new methods. 

God job!

**Pull request merged**